From f285826766bdccddf9ccdabb3f47a3552eafebdb Mon Sep 17 00:00:00 2001
From: Milad Fa <mfarazma@ibm.com>
Date: Wed, 11 Mar 2026 19:54:08 +0000
Subject: [PATCH] PPC/s390: [wasm] Fix walking WasmLiftoffSetupFrame

Port dec5f6bdbe683eb74f7908310ac9a742d057181a

Original Commit Message:

    It can happen that a Liftoff code object that just called its own
    frame setup has immediately afterwards been flushed due to memory
    pressure. In that case, we can't look it up by func index via the
    NativeModule (because that would return nullptr); instead we must
    look it up by contained PC address via the WasmCodeManager.

R=jkummerow@chromium.org, junyan1@ibm.com
BUG=
LOG=N

Change-Id: I9224d002f0b963db320333516b1c75194dadf976
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/7658728
Commit-Queue: Milad Farazmand <mfarazma@ibm.com>
Reviewed-by: Jakob Kummerow <jkummerow@chromium.org>
Reviewed-by: John <junyan1@ibm.com>
Cr-Commit-Position: refs/heads/main@{#105745}
---
 src/execution/ppc/frame-constants-ppc.h   | 1 +
 src/execution/s390/frame-constants-s390.h | 1 +
 2 files changed, 2 insertions(+)

diff --git a/src/3rdparty/chromium/v8/src/execution/ppc/frame-constants-ppc.h b/src/3rdparty/chromium/v8/src/execution/ppc/frame-constants-ppc.h
index 7de5de291901..572b8b95ec3c 100644
--- a/src/3rdparty/chromium/v8/src/execution/ppc/frame-constants-ppc.h
+++ b/src/3rdparty/chromium/v8/src/execution/ppc/frame-constants-ppc.h
@@ -33,6 +33,7 @@ class WasmLiftoffSetupFrameConstants : public TypedFrameConstants {
   static constexpr int kNumberOfSavedFpParamRegs = 8;
 
   // There's one spilled value (which doesn't need visiting) below the instance.
+  static constexpr int kCallingPCOffset = TYPED_FRAME_PUSHED_VALUE_OFFSET(0);
   static constexpr int kInstanceSpillOffset =
       TYPED_FRAME_PUSHED_VALUE_OFFSET(1);
 
diff --git a/src/3rdparty/chromium/v8/src/execution/s390/frame-constants-s390.h b/src/3rdparty/chromium/v8/src/execution/s390/frame-constants-s390.h
index f30ecac104a3..0718b2d6598f 100644
--- a/src/3rdparty/chromium/v8/src/execution/s390/frame-constants-s390.h
+++ b/src/3rdparty/chromium/v8/src/execution/s390/frame-constants-s390.h
@@ -33,6 +33,7 @@ class WasmLiftoffSetupFrameConstants : public TypedFrameConstants {
   static constexpr int kNumberOfSavedFpParamRegs = 4;
 
   // There's one spilled value (which doesn't need visiting) below the instance.
+  static constexpr int kCallingPCOffset = TYPED_FRAME_PUSHED_VALUE_OFFSET(0);
   static constexpr int kInstanceSpillOffset =
       TYPED_FRAME_PUSHED_VALUE_OFFSET(1);
 
